python/ray/serve/_private/router.py

```diff
@@ -182,6 +187,33 @@ def _try_assign_replica(self, query: Query) -> Optional[ray.ObjectRef]:
                 )
                 self.in_flight_queries[replica].add(tracker_ref)
             return user_ref
+
+        # Try to find a replica that can handle this query.
+        # If model id is not specified, we can assign it to any non-overladed replica.
+        # If model id is specified, we can try to assign it to a replica that has
+        # loaded, and if there is no such replica, we can assign it to any
+        # non-overloaded replica.
+        if query.metadata.model_id:
+            # Try to find all replicas that has loaded the model and is not overloaded.
+            candidate_replicas = [
+                replica
+                for replica in self.in_flight_queries.keys()
+                if query.metadata.model_id in replica.model_ids
+                and len(self.in_flight_queries[replica])
+                < replica.max_concurrent_queries
+            ]
+            for replica in self.in_flight_queries.keys():
+                logger.info(f"router debug: {replica.replica_tag}, {replica.model_ids}")
+            if candidate_replicas:
+                # If there are such replicas, randomly pick one.
+                replica = random.choice(candidate_replicas)
```

Comment:
Why do we do random choice instead of round robin across the candidate replicas?